File: frontend/src/services/notificationService.ts
import apiClient from '@/lib/api';

export interface Notification {
  id: string;
  user_id: string;
  product_id: string | null;
  title: string;
  message: string;
  type: 'info' | 'success' | 'warning' | 'error';
  priority: 'low' | 'medium' | 'high' | 'urgent';
  status: 'unread' | 'read' | 'archived' | 'deleted';
  category: string;
  action_url: string | null;
  created_at: string;
  read_at: string | null;
  updated_at: string;
}

export interface NotificationListParams {
  skip?: number;
  limit?: number;
  status?: 'unread' | 'read' | 'archived' | 'deleted';
  category?: string;
  unread_only?: boolean;
}

export interface UnreadCountResponse {
  count: number;
}

export interface MarkAllReadResponse {
  updated: number;
}

export const notificationService = {
  /**
   * Get user's notifications with optional filtering
   */
  async getNotifications(
    params?: NotificationListParams
  ): Promise<Notification[]> {
    const response = await apiClient.get('/api/v1/notifications', { params });
    return response.data;
  },

  /**
   * Get a single notification by ID
   */
  async getNotification(id: string): Promise<Notification> {
    const response = await apiClient.get(`/api/v1/notifications/${id}`);
    return response.data;
  },

  /**
   * Get count of unread notifications
   */
  async getUnreadCount(): Promise<number> {
    const response = await apiClient.get<UnreadCountResponse>(
      '/api/v1/notifications/unread-count'
    );
    return response.data.count;
  },

  /**
   * Mark a notification as read
   */
  async markAsRead(id: string): Promise<Notification> {
    const response = await apiClient.patch(`/api/v1/notifications/${id}`, {
      status: 'read',
    });
    return response.data;
  },

  /**
   * Archive a notification
   */
  async archiveNotification(id: string): Promise<Notification> {
    const response = await apiClient.patch(`/api/v1/notifications/${id}`, {
      status: 'archived',
    });
    return response.data;
  },

  /**
   * Mark all notifications as read
   */
  async markAllAsRead(): Promise<number> {
    const response = await apiClient.post<MarkAllReadResponse>(
      '/api/v1/notifications/mark-all-read'
    );
    return response.data.updated;
  },

  /**
   * Delete a notification
   */
  async deleteNotification(id: string): Promise<void> {
    await apiClient.delete(`/api/v1/notifications/${id}`);
  },

  /**
   * Get recent notifications (last 10 unread)
   */
  async getRecentNotifications(): Promise<Notification[]> {
    return this.getNotifications({
      limit: 10,
      unread_only: true,
    });
  },
};

File: backend/src/services/notification_service.py
"""Notification service for creating and sending notifications."""

import logging

from sqlalchemy import func, select
from sqlalchemy.ext.asyncio import AsyncSession

from notification.models import Notification
from products.models import Product
from users.models import User, UserSettings

logger = logging.getLogger(__name__)


class NotificationService:
    """Service for managing notifications."""

    NOTIFICATION_TYPES = {
        "price_change": "Price Change Alert",
        "bsr_change": "BSR Change Alert",
        "stock_change": "Stock Status Alert",
        "review_change": "Review Change Alert",
        "system": "System Notification",
        "daily_summary": "Daily Summary",
        "weekly_report": "Weekly Report",
    }

    @classmethod
    async def create_price_change_notification(
        cls,
        db: AsyncSession,
        user: User,
        product: Product,
        old_price: float,
        new_price: float,
        threshold: float,
    ) -> Notification | None:
        """Create a price change notification.

        Args:
            user: User to notify
            product: Product that changed
            old_price: Previous price
            new_price: Current price
            threshold: Threshold percentage

        Returns:
            Notification instance or None if user has disabled this notification type
        """
        # Check user settings
        stmt = select(UserSettings).where(UserSettings.user_id == user.id)
        result = await db.execute(stmt)
        settings = result.scalar_one_or_none()

        # Create default settings if not exists
        if not settings:
            settings = UserSettings(user_id=user.id)
            db.add(settings)
            await db.commit()
            await db.refresh(settings)

        if not settings.email_notifications_enabled or not settings.price_alert_emails:
            return None

        # Calculate percentage change
        price_change = ((new_price - old_price) / old_price) * 100
        change_direction = "increased" if new_price > old_price else "decreased"

        title = f"Price Alert: {product.title[:50]}..."
        message = (
            f"The price of {product.title} has {change_direction} by {abs(price_change):.1f}%.\n\n"
            f"Previous price: ${old_price:.2f}\n"
            f"Current price: ${new_price:.2f}\n"
            f"Change: ${abs(new_price - old_price):.2f}"
        )

        notification = Notification(
            user_id=user.id,
            notification_type="price_change",
            title=title,
            message=message,
            product_id=product.id,
            data={
                "old_price": old_price,
                "new_price": new_price,
                "change_percentage": price_change,
                "threshold": threshold,
            },
            priority="high" if abs(price_change) > threshold * 2 else "normal",
            action_url=f"/products/{product.id}",
        )
        db.add(notification)
        await db.commit()
        await db.refresh(notification)

        # TODO: Queue email sending task
        # await email_service.send_notification_email(user, notification)

        return notification

    @classmethod
    async def create_bsr_change_notification(
        cls,
        db: AsyncSession,
        user: User,
        product: Product,
        old_bsr: int,
        new_bsr: int,
        threshold: float,
    ) -> Notification | None:
        """Create a BSR change notification."""
        stmt = select(UserSettings).where(UserSettings.user_id == user.id)
        result = await db.execute(stmt)
        settings = result.scalar_one_or_none()

        if not settings:
            settings = UserSettings(user_id=user.id)
            db.add(settings)
            await db.commit()
            await db.refresh(settings)

        if not settings.email_notifications_enabled or not settings.bsr_alert_emails:
            return None

        # Calculate percentage change (lower BSR is better)
        bsr_change = ((new_bsr - old_bsr) / old_bsr) * 100
        change_direction = "improved" if new_bsr < old_bsr else "declined"

        title = f"BSR Alert: {product.title[:50]}..."
        message = (
            f"The Best Seller Rank of {product.title} has {change_direction} by {abs(bsr_change):.1f}%.\n\n"
            f"Previous rank: #{old_bsr:,}\n"
            f"Current rank: #{new_bsr:,}\n"
            f"Change: {abs(new_bsr - old_bsr):,} positions"
        )

        notification = Notification(
            user_id=user.id,
            notification_type="bsr_change",
            title=title,
            message=message,
            product_id=product.id,
            data={
                "old_bsr": old_bsr,
                "new_bsr": new_bsr,
                "change_percentage": bsr_change,
                "threshold": threshold,
            },
            priority="high" if abs(bsr_change) > threshold * 2 else "normal",
            action_url=f"/products/{product.id}",
        )
        db.add(notification)
        await db.commit()
        await db.refresh(notification)

        return notification

    @classmethod
    async def create_stock_change_notification(
        cls,
        db: AsyncSession,
        user: User,
        product: Product,
        old_status: str,
        new_status: str,
    ) -> Notification | None:
        """Create a stock status change notification."""
        stmt = select(UserSettings).where(UserSettings.user_id == user.id)
        result = await db.execute(stmt)
        settings = result.scalar_one_or_none()

        if not settings:
            settings = UserSettings(user_id=user.id)
            db.add(settings)
            await db.commit()
            await db.refresh(settings)

        if not settings.email_notifications_enabled or not settings.stock_alert_emails:
            return None

        title = f"Stock Alert: {product.title[:50]}..."
        message = (
            f"The stock status of {product.title} has changed.\n\n"
            f"Previous status: {old_status}\n"
            f"Current status: {new_status}"
        )

        # Prioritize "out of stock" alerts
        priority = "urgent" if "out of stock" in new_status.lower() else "normal"

        notification = Notification(
            user_id=user.id,
            notification_type="stock_change",
            title=title,
            message=message,
            product_id=product.id,
            data={
                "old_status": old_status,
                "new_status": new_status,
            },
            priority=priority,
            action_url=f"/products/{product.id}",
        )
        db.add(notification)
        await db.commit()
        await db.refresh(notification)

        return notification

    @classmethod
    async def create_system_notification(
        cls,
        db: AsyncSession,
        user: User,
        title: str,
        message: str,
        priority: str = "normal",
        action_url: str | None = None,
    ) -> Notification:
        """Create a system notification."""
        notification = Notification(
            user_id=user.id,
            notification_type="system",
            title=title,
            message=message,
            priority=priority,
            action_url=action_url,
        )
        db.add(notification)
        await db.commit()
        await db.refresh(notification)

        return notification

    @classmethod
    async def get_unread_count(cls, db: AsyncSession, user: User) -> int:
        """Get count of unread notifications for a user."""
        stmt = (
            select(func.count())
            .select_from(Notification)
            .where(Notification.user_id == user.id, Notification.is_read.is_(False))
        )
        result = await db.execute(stmt)
        count = result.scalar_one()
        return count

    @classmethod
    async def mark_all_as_read(cls, db: AsyncSession, user: User) -> int:
        """Mark all notifications as read for a user.

        Returns:
            Number of notifications marked as read
        """
        from datetime import datetime

        from sqlalchemy import update

        stmt = (
            update(Notification)
            .where(Notification.user_id == user.id, Notification.is_read.is_(False))
            .values(is_read=True, read_at=datetime.utcnow())
        )
        result = await db.execute(stmt)
        await db.commit()
        return int(result.rowcount)  # type: ignore[attr-defined]

    @classmethod
    async def delete_old_notifications(cls, db: AsyncSession, days: int = 30) -> int:
        """Delete notifications older than specified days.

        Args:
            days: Number of days to keep notifications

        Returns:
            Number of notifications deleted
        """
        from datetime import datetime, timedelta

        from sqlalchemy import delete

        cutoff_date = datetime.utcnow() - timedelta(days=days)
        stmt = delete(Notification).where(Notification.created_at < cutoff_date)
        result = await db.execute(stmt)
        await db.commit()
        return int(result.rowcount)  # type: ignore[attr-defined]
